File: src/template.cpp
#include <template.hpp>

void topic_callback(const package_template::TemplateMessage::ConstPtr& msg) {
    ROS_INFO("%s", msg->Text.c_str());
}

bool service_callback(package_template::TemplateService::Request &req, package_template::TemplateService::Response &res) {
    ROS_INFO("%s", req.Request.c_str());
    res.Response = "Response from C++";
    return true;
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "package_template_cpp");

    ros::NodeHandle nh;

    subscriber = nh.subscribe("/cpp_subscriber", 1, topic_callback);
    publisher = nh.advertise<package_template::TemplateMessage>("/python_subscriber", 1);
    service = nh.advertiseService("/cpp_service", service_callback);
    client = nh.serviceClient<package_template::TemplateService>("/python_service");

    ros::Rate rate(10);  
    while (ros::ok()) {
        package_template::TemplateMessage msg;
        msg.Text = "Hello from C++";
        publisher.publish(msg);

        package_template::TemplateService srv;
        srv.request.Request = "Request from C++";
        client.call(srv);

        ros::spinOnce();
        rate.sleep();
    }

    return 0;
}

File: scripts/template.py
#!/usr/bin/env python

import rospy

from package_template.msg import TemplateMessage
from package_template.srv import TemplateService, TemplateServiceResponse

def service_callback(req):
    rospy.loginfo(req.Request)
    return TemplateServiceResponse(Response="Response from Python")

def topic_callback(msg):
    rospy.loginfo(msg.Text)

def main():
    rospy.init_node('package_template_python')

    subscriber = rospy.Subscriber("/python_subscriber", TemplateMessage, topic_callback)
    publisher = rospy.Publisher('/cpp_subscriber', TemplateMessage, queue_size=10)
    server = rospy.Service('/python_service', TemplateService, service_callback)
    client = rospy.ServiceProxy('/cpp_service', TemplateService)

    rate = rospy.Rate(10)
    while not rospy.is_shutdown():
        message = TemplateMessage()
        message.Text = "Hello from Python"
        publisher.publish(message)

        rospy.wait_for_service('/cpp_service')
        response = client("Request from Python")

        rate.sleep()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass